src/Dscp/Educator/Web/Student/Types.hs

```diff
@@ -0,0 +1,73 @@
+
+-- | Datatypes for Student HTTP API
+
+module Dscp.Educator.Web.Student.Types
+       ( Course (..)
+       , Assignment (..)
+       , Submission (..)
+       , Grade (..)
+       , BlkProof (..)
+       , ErrResponse (..)
+       ) where
+
+import Data.Aeson.Options (defaultOptions)
+import Data.Aeson.TH (deriveJSON)
+import Data.Time.Clock (UTCTime)
+
+import Dscp.Core.Aeson ()
+import qualified Dscp.Core.Types as Core
+import Dscp.Crypto (Hash, Raw)
+import Dscp.Educator.Txs (PrivateTx (..))
+import Dscp.Educator.Web.Student.Error (APIError)
+import Dscp.Util.Aeson (AsByteString, Base64Encoded)
+
+data Course = Course
+    { cId         :: !Core.Course
+    , cDesc       :: !Text
+    , cSubjects   :: ![Core.Subject]
+    , cIsEnrolled :: !Bool
+    } deriving (Show, Eq, Generic)
+
+data Assignment = Assignment
+    { aHash           :: !(Hash Core.Assignment)
+    , aCourseId       :: !Core.Course
+    , aContentsHash   :: !(Hash Raw)
+    , aIsFinal        :: !Bool
+    , aDesc           :: !Text
+    , aLastSubmission :: !(Maybe Submission)
+    } deriving (Show, Eq, Generic)
+
+data Submission = Submission
+    { sHash           :: !(Hash Core.Submission)
+    , sContentsHash   :: !(Hash Raw)
+    , sAssignmentHash :: !(Hash Core.Assignment)
+    , sGrade          :: !(Maybe Grade)
+    } deriving (Show, Eq, Generic)
```

Comment:
I guess `sGrade` will be Nothing for request and Just for response. Won't we wish separate datatypes?

Reply:
You got it wrong. Student will send actual `Core.SignedSubmission` datatype to POST method. This adapter `Submission` type is only for responses. A submission may be already graded or not graded yet, hence `Maybe` here.